### ISSUE TITLE
as.section(argo) can be very slow

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#1120]
TL;DR: `as.section(argo)` can be slow for large files, so I speedered it up.

## Long version

In the course of playing around with the data file from Issue #1118, I discovered that converting the `argo` object to a `section` object (e.g. to make pretty "section"-type plots was *extremely* slow). A `system.time()` on my machine gave:
```
> system.time(s <- as.section(d)) # using develop
   user  system elapsed 
121.920   0.064 122.126 
```
To see what was taking so long, I profiled the call, using:
```r
tmp <- tempfile()
Rprof(tmp)
s <- as.section(d)
Rprof()
summaryRprof(tmp)
```
to get:
```
$by.self
                     self.time self.pct total.time total.pct
"=="                    119.40    97.82     119.40     97.82
"as.vector"               0.96     0.79       0.96      0.79
"as.ctd"                  0.82     0.67       2.04      1.67
"as.character"            0.38     0.31       0.38      0.31
"inherits"                0.20     0.16       0.20      0.16
"as.section"              0.12     0.10     122.02     99.97
"[["                      0.02     0.02       0.04      0.03
".local"                  0.02     0.02       0.04      0.03
"@<-"                     0.02     0.02       0.02      0.02
".getClassFromCache"      0.02     0.02       0.02      0.02
"%in%"                    0.02     0.02       0.02      0.02
"ls"                      0.02     0.02       0.02      0.02
"match"                   0.02     0.02       0.02      0.02
"substring"               0.02     0.02       0.02      0.02
"unique.default"          0.02     0.02       0.02      0.02

$by.total
                               total.time total.pct self.time self.pct
"as.section"                       122.02     99.97      0.12     0.10
"=="                               119.40     97.82    119.40    97.82
"as.ctd"                             2.04      1.67      0.82     0.67
"as.vector"                          0.96      0.79      0.96     0.79
"factor"                             0.42      0.34      0.00     0.00
"as.character"                       0.38      0.31      0.38     0.31
"inherits"                           0.20      0.16      0.20     0.16
"<Anonymous>"                        0.06      0.05      0.00     0.00
"FUN"                                0.06      0.05      0.00     0.00
"lapply"                             0.06      0.05      0.00     0.00
"[["                                 0.04      0.03      0.02     0.02
".local"                             0.04      0.03      0.02     0.02
"initialize"                         0.04      0.03      0.00     0.00
"new"                                0.04      0.03      0.00     0.00
"Reduce"                             0.04      0.03      0.00     0.00
".rs.getCompletionsSearchPath"       0.04      0.03      0.00     0.00
"unlist"                             0.04      0.03      0.00     0.00
"@<-"                                0.02      0.02      0.02     0.02
".getClassFromCache"                 0.02      0.02      0.02     0.02
"%in%"                               0.02      0.02      0.02     0.02
"ls"                                 0.02      0.02      0.02     0.02
"match"                              0.02      0.02      0.02     0.02
"substring"                          0.02      0.02      0.02     0.02
"unique.default"                     0.02      0.02      0.02     0.02
"callNextMethod"                     0.02      0.02      0.00     0.00
"getClass"                           0.02      0.02      0.00     0.00
".nextMethod"                        0.02      0.02      0.00     0.00
".rs.fuzzyMatches"                   0.02      0.02      0.00     0.00
".rs.objectsOnSearchPath"            0.02      0.02      0.00     0.00
".rs.startsWith"                     0.02      0.02      0.00     0.00
"unique"                             0.02      0.02      0.00     0.00

$sample.interval
[1] 0.02

$sampling.time
[1] 122.06
```
Note that most of the time is being spent in `==`. Looking at the code, it was clear that the offendingly slow function call was around Line 2664:
```r
        for (i in 1:nstation) {
            ##message("ARGO CASE. i: ", i, ", name:", stationLevels[i])
            look <- station==stationLevels[i]
            ctds[[i]] <- as.ctd(salinity[look], temperature[look], pressure[look],
                                longitude=longitude[look][1], latitude=latitude[look][1],
                                startTime=as.POSIXct(time[i]), station=paste("profile", stationLevels[i]))
```
I could see this was because the original matrices from the argo object were being converted to vectors, and then parsed up into `ctd` objects by doing the logical comparison of the `stationLevels` vector. For that file, the vector length ends up being about 1.5e6, and with 250 odd profiles it can really slow down.

So, in the `cr-section` branch, I re-factored the ctd-splitting to just pull the ctd data straight from the columns of the argo matrix. I'll make a PR once I'm done typing this, but the essence is to do:
```r
        for (i in 1:nstation) {
            ctds[[i]] <- as.ctd(salinity[,i], temperature[,i], pressure[,i],
                                longitude=longitude[i], latitude=latitude[i],
                                startTime=as.POSIXct(time[i]), station=as.character(station[i]))
```
where `salinity` etc are now just matrices. The speedup is significant:
```r
> system.time(s <- as.section(d)) # using cr-section
   user  system elapsed 
   0.24    0.00    0.24 
```

Q for @dankelley: Can you see any reason why doing the splitting on a vector is better than pulling rows from the matrix (I ask because it was you who coded it that way originally)?

**Comment by richardsc** [comment]
See PR #1121 .

And a nice plot!:
```r
library(oce)
d <- read.oce('nodc_6901486_prof.nc')
s <- as.section(d)
plot(s, which='temperature', xtype='time', ztype='image')
```
![image](https://cloud.githubusercontent.com/assets/233584/20374121/bdc72b8c-ac4c-11e6-8de0-981b998a98d3.png)

**Comment by richardsc** [comment]
And ooooo look, now I see the units on the section plot label like @clayton was asking for!

**Comment by dankelley** [comment]
I wonder, did you add the units or did I? I can't keep track of things. Anyway, I figure this should be closed so I'm doing that, for clarity on the issue opening page
